mirrord/layer/src/lib.rs

```diff
@@ -935,6 +935,33 @@ pub(crate) fn close_layer_fd(fd: c_int) {
     // Remove from sockets, also removing the `ConnectionQueue` associated with the socket.
     if let Some((_, socket)) = SOCKETS.remove(&fd) {
         CONNECTION_QUEUE.remove(socket.id);
+
+        // Closed file is a socket, so if it's already bound to a port - notify agent to stop
+        // mirroring/stealing that port.
+        if let Some(port) = socket.get_port() {
+            match socket.kind {
+                SocketKind::Tcp(_) => {
+                    match socket.state {
+                        // if the socket state is connected, then closing its fd is closing the
+                        // one connection, not the whole listening socket. We will know about this
+                        // close from the mirrord layer's side of the socket and will inform the
+                        // agent about the closed connection from there.
+                        SocketState::Connected(_) => {}
+                        _ => {
+                            // Ignoring errors here. We continue running, potentially without
+                            // informing the layer's and agent's TCP handlers about the socket
+                            // close. The agent might try to continue sending incoming
+                            // connections/data.
+                            let _ = blocking_send_hook_message(HookMessage::Tcp(
+                                tcp::TcpIncoming::Close(port),
+                            ));
+                        }
+                    }
+                }
+                // We don't do incoming UDP, so no need to notify anyone about this.
+                SocketKind::Udp(_) => {}
+            }
+        }
```

Comment:
Add this whole thing as a method for `UserSocket`, and here just call:

```suggestion
        socket.close();
```

Reply:
Done here https://github.com/metalbear-co/mirrord/pull/1557/commits/e5e6226e86c6fd8d5438197de07c9356c194288f. Also changed `get_port` to `get_bound_port` (return `None` if socket state is `Connected`), so that we don't match `socket.state` twice.